orderer/common/follower/follower_chain.go

```diff
@@ -527,11 +523,24 @@ func (c *Chain) pullUntilTarget(targetHeight uint64, updateEndpoints bool) (uint
 			if nextBlock == nil {
 				return n, errors.WithMessagef(cluster.ErrRetryCountExhausted, "failed to pull block %d", seq)
 			}
+
 			reportedPrevHash := nextBlock.Header.PreviousHash
 			if (nextBlock.Header.Number > 0) && !bytes.Equal(reportedPrevHash, actualPrevHash) {
-				return n, errors.Errorf("block header mismatch on sequence %d, expected %x, got %x",
+				return n, errors.Errorf("block header previous hash mismatch on sequence %d, expected %x, got %x",
 					nextBlock.Header.Number, actualPrevHash, reportedPrevHash)
 			}
+
+			if c.joinBlock != nil && c.joinBlock.Header.Number == nextBlock.Header.Number {
+				if !proto.Equal(nextBlock.Header, c.joinBlock.Header) {
+					c.logger.Errorf("Block header mismatch between the block we pulled and the block we joined with, sequence %d", c.joinBlock.Header.Number)
+					return n, errors.Errorf("block header mismatch between the block we pulled and the block we joined with, sequence %d", c.joinBlock.Header.Number)
+				}
+				if !proto.Equal(nextBlock.Data, c.joinBlock.Data) {
```

Comment:
why do we need the data check? The block puller assures us the data is the hash in the header, no?

Reply:
or are you trying to catch a case if we got a join block with a header hash that mismatches its data? if that's the reason then we should check it earlier and not here.

Reply:
That's what I'm saying  - we should verify the data matches the hash when we digest the block at the time of processing the API call, not **here**.

Reply:
Done